pkg/engine/plugins.go

```diff
@@ -262,25 +270,79 @@ func installPlugin(ctx context.Context, plugin workspace.PluginSpec) error {
 	logging.V(preparePluginVerboseLog).Infof(
 		"installPlugin(%s, %s): initiating download", plugin.Name, plugin.Version)
 
-	withProgress := func(stream io.ReadCloser, size int64) io.ReadCloser {
-		return workspace.ReadCloserProgressBar(stream, size, "Downloading plugin", cmdutil.GetGlobalColorization())
+	pluginID := fmt.Sprintf("%s-%s", plugin.Name, plugin.Version)
+	downloadMessage := "Downloading plugin " + pluginID
+
+	// Where possible, we'll use a progress bar to show download progress by
+	// wrapping the download stream with a progress reporting ReadCloser.
+	var withDownloadProgress func(io.ReadCloser, int64) io.ReadCloser
+	if opts == nil {
+		withDownloadProgress = func(stream io.ReadCloser, size int64) io.ReadCloser {
+			return workspace.ReadCloserProgressBar(
+				stream,
+				size,
+				downloadMessage,
+				cmdutil.GetGlobalColorization(),
+			)
+		}
+	} else {
+		withDownloadProgress = func(stream io.ReadCloser, size int64) io.ReadCloser {
+			return NewProgressReportingCloser(
+				opts.Events,
+				PluginDownload,
+				string(PluginDownload)+":"+pluginID,
+				downloadMessage,
+				size,
+				100*time.Millisecond, /*reportingInterval */
+				stream,
+			)
+		}
 	}
 	retry := func(err error, attempt int, limit int, delay time.Duration) {
 		logging.V(preparePluginVerboseLog).Infof(
 			"Error downloading plugin: %s\nWill retry in %v [%d/%d]", err, delay, attempt, limit)
 	}
 
-	tarball, err := workspace.DownloadToFile(plugin, withProgress, retry)
+	tarball, err := workspace.DownloadToFile(plugin, withDownloadProgress, retry)
 	if err != nil {
 		return fmt.Errorf("failed to download plugin: %s: %w", plugin, err)
 	}
 	defer func() { contract.IgnoreError(os.Remove(tarball.Name())) }()
 
-	fmt.Fprintf(os.Stderr, "[%s plugin %s-%s] installing\n", plugin.Kind, plugin.Name, plugin.Version)
-
 	logging.V(preparePluginVerboseLog).Infof(
 		"installPlugin(%s, %s): extracting tarball to installation directory", plugin.Name, plugin.Version)
-	if err := plugin.InstallWithContext(ctx, workspace.TarPlugin(tarball), false); err != nil {
+
+	// In a similar manner to downloads, we'll use a progress bar to show install
+	// progress by wrapping the download stream with a progress reporting
+	// ReadCloser where possible.
+	var withInstallProgress func(io.ReadCloser) io.ReadCloser
+	stat, err := tarball.Stat()
+	if opts == nil || err != nil {
+		withInstallProgress = func(stream io.ReadCloser) io.ReadCloser {
+			return stream
+		}
+
+		installing := fmt.Sprintf("[%s plugin %s] installing", plugin.Kind, pluginID)
+		fmt.Fprintln(os.Stderr, installing)
```

Comment:
Do we still want this message? When is this shown?

Reply:
The `isInteractive` check is here https://github.com/pulumi/pulumi/pull/17019/files#diff-243d1693b6c2c0cf3a92f1b238580dce7620becea7f70edc67d52f4de318b60aR244.

The message above is shown in the case that we don't have a set of deployment options (and thus an event emitter to use to send messages along). I could have sworn that there was a set of cases where this happens (e.g. we run an explicit install), but am now struggling to find this code. Perhaps I imagined it, or it was recently refactored, but I suspect I can refine this case now possibly.

Reply:
> The `isInteractive` check is here https://github.com/pulumi/pulumi/pull/17019/files#diff-243d1693b6c2c0cf3a92f1b238580dce7620becea7f70edc67d52f4de318b60aR244.

Ah is `render` only ever called when we're interactive?